core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java

```diff
@@ -102,8 +102,15 @@ public boolean processExistingRecoveries(RoutingAllocation allocation) {
                 if (matchingNodes.getNodeWithHighestMatch() != null) {
                     DiscoveryNode currentNode = allocation.nodes().get(shard.currentNodeId());
                     DiscoveryNode nodeWithHighestMatch = matchingNodes.getNodeWithHighestMatch();
+                    // current node will not be in matchingNodes as it is filtered away by SameShardAllocationDecider
+                    final String currentSyncId;
+                    if (shardStores.getData().containsKey(currentNode)) {
+                        currentSyncId = shardStores.getData().get(currentNode).storeFilesMetaData().syncId();
+                    } else {
+                        currentSyncId = null;
+                    }
                     if (currentNode.equals(nodeWithHighestMatch) == false
```

Comment:
The current node can never be `nodeWithHighestMatch` as `findMatchingNodes` will not add current node due to `SameShardAllocationDecider`. This means that `currentNode.equals(nodeWithHighestMatch) == false` always holds and can thus be removed from the condition (or you can add it as assertion).